File: php/src/Validate/Classification.php
<?php

declare(strict_types=1);

// SPDX-License-Identifier: MIT

namespace HopTop\Vstar\Validate;

use HopTop\Vstar\Component;
use HopTop\Vstar\Generated\Codes;
use HopTop\Vstar\Transp;
use HopTop\Vstar\VClass;

/**
 * spec/05 §8 -- the CLASS and TRANSP value domains.
 *
 * @internal
 */
final class Classification
{
    private const PROP_CLASS = 'CLASS';

    private const PROP_TRANSP = 'TRANSP';

    /**
     * Flag a CLASS outside the RFC 5545 §3.8.1.3 vocabulary and a TRANSP
     * outside §3.8.2.7's.
     *
     * Comparison is case-insensitive (spec/05 §8, RFC 5545 §3.1). The
     * value is checked on any component carrying the property -- there is
     * no type gating, because V* diagnoses no scope rule for any property.
     * An X- or IANA token on CLASS, which the RFC's ABNF admits, is still
     * flagged: spec/05 §8 binds the value to the three registered names.
     *
     * @return list<Diagnostic>
     */
    public static function check(Component $c, string $path): array
    {
        $out = [];

        $d = self::vocabularyDiagnostic(
            $c,
            $path,
            self::PROP_CLASS,
            self::classVocabulary(),
            Codes::CLASS_NOT_IN_VOCABULARY,
            'RFC 5545 §3.8.1.3',
        );

        if ($d !== null) {
            $out[] = $d;
        }

        $d = self::vocabularyDiagnostic(
            $c,
            $path,
            self::PROP_TRANSP,
            self::transpVocabulary(),
            Codes::TRANSP_NOT_IN_VOCABULARY,
            'RFC 5545 §3.8.2.7',
        );

        if ($d !== null) {
            $out[] = $d;
        }

        return $out;
    }

    /**
     * The STATUS-shaped diagnostic for property `$name` on `$c` when its
     * value is outside `$allowed`, or null when the property is absent or
     * its value is allowed.
     *
     * @param list<string> $allowed
     */
    private static function vocabularyDiagnostic(
        Component $c,
        string $path,
        string $name,
        array $allowed,
        string $code,
        string $ref,
    ): ?Diagnostic {
        $p = $c->get($name);

        if ($p === null) {
            return null;
        }

        foreach ($allowed as $want) {
            if (Internal::equalFold($p->value, $want)) {
                return null;
            }
        }

        return Diagnostic::of(
            $code,
            $name . ' value ' . $p->value . ' is not valid; allowed: '
                . implode(', ', $allowed) . ' (' . $ref . ')',
            $path . '.' . $name,
        );
    }

    /**
     * The CLASS values RFC 5545 §3.8.1.3 admits, projected from the
     * port's own backed enum rather than read out of
     * `Codes::CLASS_VOCABULARY`, for the reason {@see Status} gives: these
     * are the values the codec encodes against, so the table cannot
     * disagree with what this library writes. The registry's copy is
     * reconciled against the enum in
     * `tests/Validate/RegistryVocabularyTest.php`.
     *
     * @return list<string>
     */
    private static function classVocabulary(): array
    {
        return [
            VClass::Public->value,
            VClass::Private->value,
            VClass::Confidential->value,
        ];
    }

    /**
     * The TRANSP values RFC 5545 §3.8.2.7 admits, from the backed enum for
     * the same reason as {@see self::classVocabulary()}.
     *
     * @return list<string>
     */
    private static function transpVocabulary(): array
    {
        return [
            Transp::Opaque->value,
            Transp::Transparent->value,
        ];
    }
}

File: rs/src/validate/classification.rs
// SPDX-License-Identifier: MIT

//! spec/05 §8 — the `CLASS` and `TRANSP` value domains.

use super::{diagnostic, Diagnostic};
use crate::generated::codes::{CLASS_NOT_IN_VOCABULARY, TRANSP_NOT_IN_VOCABULARY};
use crate::{Class, Component, Transp};

/// The closed-vocabulary property names this rule checks.
const PROP_CLASS: &str = "CLASS";
const PROP_TRANSP: &str = "TRANSP";

/// The `CLASS` values RFC 5545 §3.8.1.3 allows.
///
/// Built from the crate's own wire enum rather than the generated
/// [`CLASS_VOCABULARY`](crate::generated::codes::CLASS_VOCABULARY), for
/// the reason `status` gives: the enum is what the codec encodes
/// against, so a table built from it cannot disagree with what this
/// library writes. The registry's copy is reconciled against the enum
/// in `tests/registry_vocabulary.rs`.
fn class_vocabulary() -> Vec<&'static str> {
    Class::ALL.iter().map(Class::as_str).collect()
}

/// The `TRANSP` values RFC 5545 §3.8.2.7 allows; see
/// [`class_vocabulary`] for why the source is the enum.
fn transp_vocabulary() -> Vec<&'static str> {
    Transp::ALL.iter().map(Transp::as_str).collect()
}

/// Flag a `CLASS` outside the RFC 5545 §3.8.1.3 vocabulary and a
/// `TRANSP` outside §3.8.2.7's, one finding each.
///
/// Comparison is case-insensitive (spec/05 §8, RFC 5545 §3.1). The
/// value is checked on any component carrying the property — there is
/// no type gating, because V\* diagnoses no scope rule for any
/// property. An `X-` or IANA token on `CLASS`, which the RFC's ABNF
/// admits, is still a finding: spec/05 §8 binds the value to the three
/// registered names.
pub(super) fn check(c: &Component, path: &str) -> Vec<Diagnostic> {
    let mut out = Vec::new();
    out.extend(vocabulary_diagnostic(
        c,
        path,
        PROP_CLASS,
        &class_vocabulary(),
        CLASS_NOT_IN_VOCABULARY,
        "RFC 5545 §3.8.1.3",
    ));
    out.extend(vocabulary_diagnostic(
        c,
        path,
        PROP_TRANSP,
        &transp_vocabulary(),
        TRANSP_NOT_IN_VOCABULARY,
        "RFC 5545 §3.8.2.7",
    ));
    out
}

/// The `STATUS`-shaped diagnostic for property `name` on `c` when its
/// value is outside `allowed`; `None` when the property is absent or
/// its value is allowed.
fn vocabulary_diagnostic(
    c: &Component,
    path: &str,
    name: &str,
    allowed: &[&str],
    code: &'static str,
    reference: &str,
) -> Option<Diagnostic> {
    let p = c.get(name)?;
    if allowed
        .iter()
        .any(|want| p.value.eq_ignore_ascii_case(want))
    {
        return None;
    }
    Some(diagnostic(
        code,
        format!(
            "{name} value {} is not valid; allowed: {} ({reference})",
            p.value,
            allowed.join(", ")
        ),
        format!("{path}.{name}"),
    ))
}
